#466 TelegramLogger for SupervisedRunner

andrey-avdeev opened the pull request:
## Description

TelegramLogger was added to catalyst.dl.callbacks.logging

## Related Issue

https://github.com/catalyst-team/catalyst/issues/466

## Type of Change

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [X] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [X] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [X] I have checked the code-style using `make check-style`.
- [x] I have written the docstring in Google format for all the methods and classes that I used.
- [x] I have checked the docs using `make check-docs`.

andrey-avdeev committed:
telegram logger was added

bagxi commented on catalyst/dl/callbacks/logging.py:
```diff
@@ -183,6 +186,92 @@ def on_stage_end(self, state: RunnerState):
+
+    def __init__(
+        self,
+        token: str,
```
It is very useful to add all your experiment files into git (including configs), and at the same time, you don't really want to add any secrets into git. So, in my opinion, it is better to read `chat_id` and `token` from env variables or something like `telegram.cfg`.

andrey-avdeev replied:
I'm sure that this feature (configuration parsing) must be not it telegram logger itself. Configuration manager (like in various .net or java frameworks) or global configuration (like in Flask), or catalyst.global.config['telegram_chat_id']  - something this.

bagxi commented on catalyst/dl/callbacks/logging.py:
```diff
@@ -183,6 +186,92 @@ def on_stage_end(self, state: RunnerState):
+    def _send_text(self, text: str):
+        url = f"{self._base_url}?chat_id={self._chat_id}&disable_web_page_preview=1&text={quote_plus(text, safe='')}"
+
+        request = Request(url)
```
It seems that pipeline can fall with error in case of problems with an internet connection. Can you check if `TelegramLogger` suffers from this and add something like `try-except` if required?

andrey-avdeev replied:
done

andrey-avdeev committed:
try/except was added to telegram logger

andrey-avdeev committed:
flake8 formatting

andrey-avdeev committed:
Merge github.com:catalyst-team/catalyst

# Conflicts:
#	catalyst/dl/callbacks/logging.py

andrey-avdeev committed:
Merge github.com:catalyst-team/catalyst

# Conflicts:
#	catalyst/dl/callbacks/__init__.py

andrey-avdeev committed:
telegram logger fix

andrey-avdeev committed:
missing imports fix

andrey-avdeev committed:
Merge github.com:catalyst-team/catalyst

andrey-avdeev committed:
check docs

andrey-avdeev committed:
docs fix

andrey-avdeev committed:
merge from master with resolve conflicts

andrey-avdeev committed:
telegram pics were removed

andrey-avdeev committed:
Merge github.com:catalyst-team/catalyst

andrey-avdeev committed:
logging isosort and pre-commit check

andrey-avdeev committed:
isort check

bagxi committed:
Merge branch 'master' of github.com:catalyst-team/catalyst

bagxi committed:
add `on_exception` & fix codestyle

bagxi committed:
print -> logging warning

Scitator committed:
Update logging.py

tokens from environ to make it more user-friendly
we need global catalyst config

Scitator committed:
Update logging.py

codestyle fix

bagxi committed:
add metric formatting, small fixes

bagxi committed:
add metric formatting, small fixes

bagxi committed:
Merge branch 'master' of github.com:andrey-avdeev/catalyst

TezRomacH commented on catalyst/dl/callbacks/logging.py:
```diff
@@ -288,50 +291,61 @@ def _send_text(self, text: str):
         except Exception as e:
             logging.getLogger(__name__).warning(f"telegram.send.error:{e}")
 
+    def _format_metric(self, name: str, value: float) -> str:
```
this function appears 3 times in catalyst. Maybe is it better to move it somewhere?

bagxi replied:
Is it possible to make the `TxtMetricsFormatter._format_metric` static?

TezRomacH replied:
yes via `@staticmethod`, but it's better to move this func to utils

bagxi committed:
`_format_metric` -> `utils.misc.format_metric`

bagxi commented on catalyst/dl/utils/formatters.py:
```diff
@@ -36,29 +44,19 @@ class TxtMetricsFormatter(MetricsFormatter):
+        no need to use it directly!
     """
 
     def __init__(self):
```
What docstring can be added here?

TezRomacH replied:
> Initializes the TxtMetricsFormatter

good enough

bagxi commented on catalyst/utils/misc.py:
```diff
@@ -96,7 +104,7 @@ def maybe_recursive_call(
     recursive_kwargs=None,
-    **kwargs
+    **kwargs,
 ):
```
What docstring can be added here?

TezRomacH replied:
```
    """
    Calls the ``method`` recursively for the object_or_dict

    Args:
        object_or_dict (Any): some object or a dictinary of objects
        method (str): method name to call
        recursive_args: list of arguments to pass to the ``method``
        recursive_kwargs: list of key-arguments to pass to the ``method``
        **kwargs: Arbitrary keyword arguments
    """
```

bagxi committed:
add docstrings
